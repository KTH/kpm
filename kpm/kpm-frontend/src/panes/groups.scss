@import "../vars";

#kpm-6cf53 .kpm-groups {
  $rowGap: calc(2 * var(--kpmGap));
  $colGap: calc(4 * var(--kpmGap));
  $colCount: 3;
  $colWidth: calc(100% / $colCount - $colGap * ($colCount - 1) / $colCount);

  & > ul {
    display: flex;
    flex-flow: row wrap;
    gap: $rowGap $colGap;
    align-items: flex-start;
    width: 100%;

    & > .kpm-groups-item {
      display: flex;
      align-items: start;
      width: $colWidth;
      margin-bottom: 0;
      padding: var(--kpmPad);
      gap: $rowGap $colGap;

      .icon.star {
        color: var(--kpmIconStarBg);
        // Fix alignment error
        margin-bottom: 2px;

        &.active {
          color: var(--kpmIconStarActiveBg);
        }
      }
    }
  }
}

@media #{$mblBreakpoint} {
  #kpm-6cf53 .kpm-groups > ul > .kpm-groups-item {
    flex-basis: 100%;
    font-size: 1.1em;
  }
}
