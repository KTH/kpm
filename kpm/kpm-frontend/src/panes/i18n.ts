export type TLang = "sv" | "en";

export function i18n(
  strObj: Record<TLang, string> | string,
  lang: TLang = "sv"
) {
  if (typeof strObj === "object") {
    return strObj[lang];
  }

  // TODO: Create language dict for static string lookup
  return strObj;
}
