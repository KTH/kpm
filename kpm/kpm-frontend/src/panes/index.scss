@use "./teaching.scss";

.kpm-studies-course {
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  gap: 0.5em;

  h2,
  h3 {
    font-size: 1em;
    margin: 0;
    padding: 0;
  }
}
