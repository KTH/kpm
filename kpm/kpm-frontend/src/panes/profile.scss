@use "../vars" as *;
@use "@kth/style/scss/tokens/typography";

#kpm-6cf53 div.kpm-profile {
  h3 {
    @include typography.font-heading-s;
    margin-block-start: 1rem;
    margin-block-end: 0.25rem;
  }

  h4 {
    @include typography.font-heading-xs;
    margin-block-start: 0.25rem;
    margin-block-end: 0.125rem;
  }
}
#kpm-6cf53 .kpm-profile .kpm-profile-link {
  $rowGap: calc(2 * var(--kpmGap));
  $colGap: $rowGap;

  display: flex;
  flex-flow: row nowrap;
  gap: $rowGap $colGap;
  align-items: center;

  img {
    height: 3em;
    width: 3em;
  }
}

#kpm-6cf53 .kpm-button.change-lang::after {
  display: inline-block;
  content: " ";
  width: 1em;
  height: 1em;
  background: $data-language-black-svg;
  margin: 0.2em 0 0.2em 1ch;
}

@media #{$mblBreakpoint} {
  #kpm-6cf53 .kpm-profile {
    & > .kpm-col {
      font-size: 1.1em;
    }
  }
}
