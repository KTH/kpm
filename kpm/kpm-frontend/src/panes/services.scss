#kpm-6cf53 div.kpm-services {
  $rowGap: calc(2 * var(--kpmGap));
  $colGap: calc(4 * var(--kpmGap));
  $colCount: 3;
  $colWidth: calc(100% / $colCount - $colGap * ($colCount - 1) / $colCount);

  display: flex;
  flex-flow: row wrap;
  align-items: flex-start;
  gap: $rowGap $colGap;

  .kpm-col {
    width: $colWidth;
    flex-grow: 0;
  }

  .kpm-services-links {
    flex-grow: 2;

    & > ul {
      display: flex;
      flex-flow: row wrap;
      align-items: flex-start;
      gap: $rowGap $colGap;

      & > li {
        $colCount: 2;
        width: calc((100% - $colGap) / $colCount);
        margin-bottom: 0;
      }
    }
  }

  .kpm-empty-placeholder,
  .kpm-loading-placeholder,
  .kpm-error-message {
    min-height: 5rem;
  }
}
