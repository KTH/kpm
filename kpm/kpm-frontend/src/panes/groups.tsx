import React, { useState } from "react";
import { MenuPane, MenuPaneHeader } from "../components/menu";
import { APIGroups } from "kpm-backend-interface";
import { createApiUri, formatTerm, useDataFecther } from "./utils";
import {
  EmptyPlaceholder,
  ErrorMessage,
  LoadingPlaceholder,
} from "../components/common";
import { i18n } from "../i18n/i18n";
import "./groups.scss";
import { IconSettings, IconStar } from "../components/icons";
import { FilterOption, TabFilter } from "../components/filter";

export async function loaderStudies({ request }: any = {}): Promise<APIGroups> {
  const res = await fetch(createApiUri("/api/groups"), {
    signal: request?.signal,
  });
  const json = await res.json();
  return json;
}

type TFilter = "favs" | "all";

export function Groups() {
  const { res, loading, error } = useDataFecther<APIGroups>(loaderStudies);
  const { groups, group_search_url } = res || {};

  const [filter, setFilter] = useState<TFilter>("favs");

  const filteredGroups = groups?.filter((group) => {
    switch (filter) {
      case "favs":
        return group.starred;
      case "all":
        return true;
    }
  });

  const isEmpty = !loading && !error && filteredGroups?.length === 0;

  return (
    <MenuPane className="kpm-groups">
      <MenuPaneHeader title={i18n("My Groups")}>
        <a
          title="Search for interesting groups to follow"
          href="https://www.kth.se/search/?entityFilter=social-group&filterLabel=Groups"
        >
          {i18n("Find groups")}
        </a>
        <a href="https://www.kth.se/social/group-create/">
          {i18n("Create new group")}
        </a>
        <IconSettings href="https://www.kth.se/social/home/settings/groups" />
      </MenuPaneHeader>
      <TabFilter>
        <FilterOption<TFilter>
          value="favs"
          filter={filter}
          onSelect={setFilter}
        >
          {i18n("Favourites")}
        </FilterOption>
        <FilterOption<TFilter> value="all" filter={filter} onSelect={setFilter}>
          {i18n("All Subscriptions")}
        </FilterOption>
      </TabFilter>
      {loading && <LoadingPlaceholder />}
      {error && <ErrorMessage error={error} />}
      {isEmpty && (
        <EmptyPlaceholder>
          {filter === "favs"
            ? i18n("You have no groups marked as favourites.")
            : i18n("You don't belong to any groups.")}
        </EmptyPlaceholder>
      )}
      {!isEmpty && (
        <ul>
          {filteredGroups?.map((group) => (
            <li>
              <IconStar className={group.starred ? "star active" : "star"} />
              <a href={group.url}>{group.name}</a>
            </li>
          ))}
        </ul>
      )}
    </MenuPane>
  );
}
