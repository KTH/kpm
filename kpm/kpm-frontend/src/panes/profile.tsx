import * as React from "react";
import { useLoaderData } from "react-router-dom";
import { ErrorMessage, LoadingPlaceholder } from "../components/common";
import { MenuPane } from "../components/menu";
import { createApiUri, useDataFecther } from "./utils";
import { i18n } from "../i18n/i18n";

export async function loaderProfile({ request }: any = {}) {
  const res = await fetch("/kpm/api", {
    signal: request?.signal,
  });
  const json = await res.json();
  return json;
}

export function Profile() {
  const { res, loading, error } = useDataFecther<{ msg: string }>(
    loaderProfile
  );
  const { msg } = res || {};
  // const { msg } = useLoaderData() as { msg: string };
  return (
    <MenuPane>
      {loading && <LoadingPlaceholder />}
      {error && <ErrorMessage error={error} />}
      {msg && (
        <h2>
          {i18n("Profile")} {msg}
        </h2>
      )}
      {msg && (
        <p>
          <a href={createApiUri("/auth/logout")}>{i18n("Logout")}</a>
        </p>
      )}
    </MenuPane>
  );
}
