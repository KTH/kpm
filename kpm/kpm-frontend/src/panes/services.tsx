import React, { Fragment } from "react";
import { MenuPane, MenuPaneHeader } from "../components/menu";
import { APIServices } from "kpm-backend-interface";
import { fetchApi, useDataFecther } from "./utils";
import {
  AuthError,
  EmptyPlaceholder,
  ErrorMessage,
  LoadingPlaceholder,
} from "../components/common";
import { i18n } from "../i18n/i18n";
import "./services.scss";
import { IconSettings, IconStar } from "../components/icons";
import { FilterOption, TabFilter } from "../components/filter";

export async function loaderServices({
  request,
}: any = {}): Promise<APIServices> {
  const res = await fetchApi("/api/services", {
    signal: request?.signal,
  });
  const json = await res.json();
  if (res.ok) {
    return json;
  } else {
    if (res.status === 401) {
      throw new AuthError(json.message);
    }
    throw new Error(json.message);
  }
}

type TFilter = "favs" | "all";

export function Services() {
  const { res, loading, error } = useDataFecther<APIServices>(loaderServices);
  const { servicelinks, studentlinks } = res || {};

  const isEmptyServiceLinks = !loading && !error && servicelinks?.length === 0;
  const hasStudentlinks =
    Array.isArray(studentlinks) && studentlinks?.length > 0;
  const showStudentLinksWidget = hasStudentlinks;

  return (
    <MenuPane className="kpm-services" error={error}>
      <MenuPaneHeader title={i18n("My Services")}>
        <a
          title={i18n("Help / feedback for the personal menu")}
          href="https://www.kth.se/social/group/feedback-fran-anvand/page/personliga-menyn/"
        >
          Help / feedback
        </a>
        <IconSettings href="https://www.kth.se/social/servicelinks" />
      </MenuPaneHeader>
      <Fragment>
        <div className="kpm-col">
          <h3 className="kpm-col-header">{i18n("LADOK for Students")}</h3>
          <ul>
            <li>
              <h4>
                <a href="https://www.student.ladok.se/student/app/studentwebb/start">
                  {i18n("Home page")}
                </a>
              </h4>
              <p>
                {i18n(
                  "Relevant information right now. Course registration, Exam registration"
                )}
              </p>
            </li>
            <li>
              <h4>
                <a href="https://www.student.ladok.se/student/app/studentwebb/min-utbildning/alla">
                  {i18n("My education")}
                </a>
              </h4>
              <p>
                {i18n(
                  "Overview of your studies. Programmes, courses, results on courses"
                )}
              </p>
            </li>
            <li>
              <h4>
                <a href="https://www.student.ladok.se/student/app/studentwebb/examinationstillfallen/oppna-for-anmalan">
                  {i18n("Examinations")}
                </a>
              </h4>
              <p>{i18n("Sign up for examinations")}</p>
            </li>
            <li>
              <h4>
                <a href="https://www.student.ladok.se/student/app/studentwebb/intyg">
                  {i18n("Transcripts")}
                </a>
              </h4>
            </li>
            <li>
              <h4>
                <a href="https://www.student.ladok.se/student/app/studentwebb/examen-bevis">
                  {i18n("Degree Certificate")}
                </a>
              </h4>
              <p>{i18n("Apply for degree certificate")}</p>
            </li>
          </ul>
          {showStudentLinksWidget && (
            <Fragment>
              <h3 className="kpm-col-header">
                {i18n("Services for Students")}
              </h3>
              {loading && <LoadingPlaceholder />}
              {error && <ErrorMessage error={error} />}
              {hasStudentlinks && (
                <ul>
                  {studentlinks?.map((link) => (
                    <li key={link.url}>
                      <h4>
                        <a href={link.url}>{link.name}</a>
                      </h4>
                      {link.info && <p>{link.info}</p>}
                    </li>
                  ))}
                </ul>
              )}
            </Fragment>
          )}
        </div>
        <div className="kpm-col kpm-services-links">
          <h3 className="kpm-col-header">{i18n("Other Selected Services")}</h3>
          {loading && <LoadingPlaceholder />}
          {error && <ErrorMessage error={error} />}
          {isEmptyServiceLinks && (
            <EmptyPlaceholder>
              {i18n("You have no service links.")}
            </EmptyPlaceholder>
          )}
          {!isEmptyServiceLinks && (
            <ul>
              {servicelinks?.map((links) => (
                <li key={links.url}>
                  <h4>
                    <a href={links.url}>{links.name}</a>
                  </h4>
                </li>
              ))}
            </ul>
          )}
        </div>
      </Fragment>
    </MenuPane>
  );
}
