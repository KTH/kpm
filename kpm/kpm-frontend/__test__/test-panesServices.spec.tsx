import * as React from "react";
import { describe, expect, test, afterEach } from "@jest/globals";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { Services } from "../src/panes/services";
import { mockFetchJson, resetMockedFetch } from "fetch-for-jest";
import { APIServices } from "kpm-backend-interface";

describe("<Services />", () => {
  afterEach(() => {
    resetMockedFetch();
  });

  test("Can be rendered", () => {
    const component = renderer.create(<Services />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Can be rendered with data", async () => {
    mockFetchJson<APIServices>({
      servicelinks: [
        { name: "Test Service Link", url: "//:test/test/service" },
      ],
      studentlinks: [
        { name: "Test Student Link", url: "//:test/test/student" },
      ],
    });

    let component: any;
    await act(() => {
      component = renderer.create(<Services />);
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
