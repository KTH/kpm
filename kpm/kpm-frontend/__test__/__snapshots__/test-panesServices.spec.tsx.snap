// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Services /> Can be rendered 1`] = `
<dialog
  className="kpm-modal"
>
  <button
    className="kpm-modal-back-button kpm-mobile"
    onClick={[Function]}
  >
    Back to Personal Menu
  </button>
  <div
    className="kpm-modal-content kpm-services"
  >
    <header
      className="kpm-modal-content-header"
    >
      <h2>
        My Services
      </h2>
      <aside
        className="kpm-modal-content-header-actions"
      >
        <a
          href="https://www.kth.se/social/group/feedback-fran-anvand/page/personliga-menyn/"
          title="Help / feedback for the personal menu"
        >
          Help / feedback
        </a>
        <a
          className="icon"
          href="https://www.kth.se/social/servicelinks"
        >
          <svg
            height="1.2em"
            viewBox="0 0 86 86"
            width="1.2em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Select how you are notified, how the schedule is displayed, or remove subscriptions.
            </title>
            <path
              d="M11.41 72.18a42.98 42.98 0 0 0 20.65 12.41A10.62 10.62 0 0 1 32 83.5a10.5 10.5 0 1 1 20.91 1.35 42.95 42.95 0 0 0 21.3-12.28 10.5 10.5 0 0 1 10.3-18.3 43.05 43.05 0 0 0 0-22.54 10.5 10.5 0 0 1-10.3-18.3A42.94 42.94 0 0 0 52 .94l.01.56a10.5 10.5 0 0 1-21 .2 43 43 0 0 0-19.59 12.12A10.49 10.49 0 0 1 1.38 32.16a43.07 43.07 0 0 0 0 21.68 10.5 10.5 0 0 1 10.04 18.34zM42 57.94a15 15 0 1 0 0-30 15 15 0 0 0 0 30z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </a>
      </aside>
    </header>
    <div
      className="kpm-col"
    >
      <h3
        className="kpm-col-header"
      >
        LADOK for Students
      </h3>
      <ul>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/start"
            >
              Home page
            </a>
          </h4>
          <p>
            Relevant information right now. Course registration, Exam registration
          </p>
        </li>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/min-utbildning/alla"
            >
              My education
            </a>
          </h4>
          <p>
            Overview of your studies. Programmes, courses, results on courses
          </p>
        </li>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/examinationstillfallen/oppna-for-anmalan"
            >
              Examinations
            </a>
          </h4>
          <p>
            Sign up for examinations
          </p>
        </li>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/intyg"
            >
              Transcripts
            </a>
          </h4>
        </li>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/examen-bevis"
            >
              Degree Certificate
            </a>
          </h4>
          <p>
            Apply for degree certificate
          </p>
        </li>
      </ul>
    </div>
    <div
      className="kpm-col kpm-services-links"
    >
      <h3
        className="kpm-col-header"
      >
        Other Selected Services
      </h3>
      <div
        className="kpm-loading-placeholder"
      >
        <div
          className="kpm-loader"
        />
      </div>
      <ul />
    </div>
  </div>
</dialog>
`;

exports[`<Services /> Can be rendered with data 1`] = `
<dialog
  className="kpm-modal active"
>
  <button
    className="kpm-modal-back-button kpm-mobile"
    onClick={[Function]}
  >
    Back to Personal Menu
  </button>
  <div
    className="kpm-modal-content kpm-services"
  >
    <header
      className="kpm-modal-content-header"
    >
      <h2>
        My Services
      </h2>
      <aside
        className="kpm-modal-content-header-actions"
      >
        <a
          href="https://www.kth.se/social/group/feedback-fran-anvand/page/personliga-menyn/"
          title="Help / feedback for the personal menu"
        >
          Help / feedback
        </a>
        <a
          className="icon"
          href="https://www.kth.se/social/servicelinks"
        >
          <svg
            height="1.2em"
            viewBox="0 0 86 86"
            width="1.2em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Select how you are notified, how the schedule is displayed, or remove subscriptions.
            </title>
            <path
              d="M11.41 72.18a42.98 42.98 0 0 0 20.65 12.41A10.62 10.62 0 0 1 32 83.5a10.5 10.5 0 1 1 20.91 1.35 42.95 42.95 0 0 0 21.3-12.28 10.5 10.5 0 0 1 10.3-18.3 43.05 43.05 0 0 0 0-22.54 10.5 10.5 0 0 1-10.3-18.3A42.94 42.94 0 0 0 52 .94l.01.56a10.5 10.5 0 0 1-21 .2 43 43 0 0 0-19.59 12.12A10.49 10.49 0 0 1 1.38 32.16a43.07 43.07 0 0 0 0 21.68 10.5 10.5 0 0 1 10.04 18.34zM42 57.94a15 15 0 1 0 0-30 15 15 0 0 0 0 30z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </a>
      </aside>
    </header>
    <div
      className="kpm-col"
    >
      <h3
        className="kpm-col-header"
      >
        LADOK for Students
      </h3>
      <ul>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/start"
            >
              Home page
            </a>
          </h4>
          <p>
            Relevant information right now. Course registration, Exam registration
          </p>
        </li>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/min-utbildning/alla"
            >
              My education
            </a>
          </h4>
          <p>
            Overview of your studies. Programmes, courses, results on courses
          </p>
        </li>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/examinationstillfallen/oppna-for-anmalan"
            >
              Examinations
            </a>
          </h4>
          <p>
            Sign up for examinations
          </p>
        </li>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/intyg"
            >
              Transcripts
            </a>
          </h4>
        </li>
        <li>
          <h4>
            <a
              href="https://www.student.ladok.se/student/app/studentwebb/examen-bevis"
            >
              Degree Certificate
            </a>
          </h4>
          <p>
            Apply for degree certificate
          </p>
        </li>
      </ul>
      <h3
        className="kpm-col-header"
      >
        Services for Students
      </h3>
      <ul>
        <li>
          <h4>
            <a
              href="//:test/test/student"
            >
              Test Student Link
            </a>
          </h4>
        </li>
      </ul>
    </div>
    <div
      className="kpm-col kpm-services-links"
    >
      <h3
        className="kpm-col-header"
      >
        Other Selected Services
      </h3>
      <ul>
        <li>
          <h4>
            <a
              href="//:test/test/service"
            >
              Test Service Link
            </a>
          </h4>
        </li>
      </ul>
    </div>
  </div>
</dialog>
`;
